src/main/java/org/tron/common/overlay/portmap/router/cling/ClingPortMappingExtractor.java

```diff
@@ -0,0 +1,157 @@
+ * GNU General Public License as published by the Free Software Foundation, either version 3 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without
+ * even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License along with this program.  If
+ * not, see <http://www.gnu.org/licenses/>.
+ */
+package org.tron.common.overlay.portmap.router.cling;
+
+import java.util.Collection;
+import java.util.LinkedList;
+import org.fourthline.cling.model.message.control.IncomingActionResponseMessage;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.tron.common.overlay.portmap.model.PortMapping;
+import org.tron.common.overlay.portmap.router.RouterException;
+import org.tron.common.overlay.portmap.router.cling.action.ActionService;
+import org.tron.common.overlay.portmap.router.cling.action.GetPortMappingEntryAction;
+
+/**
+ * This class fetches all {@link PortMapping} from an {@link :InternetGatewayDevice}.
+ */
+class ClingPortMappingExtractor {
+
+  private final Logger logger = LoggerFactory.getLogger(this.getClass());
+  private final Collection<PortMapping> mappings;
+  private boolean moreEntries;
+  private int currentMappingNumber;
+
+  /**
+   * The maximum number of port mappings that we will try to retrieve from the router.
+   */
+  private final int maxNumPortMappings;
+  private final ActionService actionService;
+
+  ClingPortMappingExtractor(final ActionService actionService, final int maxNumPortMappings) {
+    this.actionService = actionService;
+    this.maxNumPortMappings = maxNumPortMappings;
+    this.mappings = new LinkedList<>();
+    this.moreEntries = true;
+    this.currentMappingNumber = 0;
+  }
+
+  public Collection<PortMapping> getPortMappings() throws RouterException {
+
+    /*
+     * This is a little trick to get all port mappings. There is a method that gets the number of available port
+     * mappings (getNatMappingsCount()), but it seems, that this method just tries to get all port mappings and
+     * checks, if an error is returned.
+     *
+     * In order to speed this up, we will do the same here, but stop, when the first exception is thrown.
+     */
+
+    while (morePortMappingsAvailable()) {
+      logger.debug("Getting port mapping with entry number " + currentMappingNumber + "...");
+
+      try {
+        final PortMapping portMapping = actionService
+            .run(new GetPortMappingEntryAction(actionService.getService(), currentMappingNumber));
+        mappings.add(portMapping);
+      } catch (final ClingOperationFailedException e) {
+        handleFailureResponse(e.getResponse());
+      }
+      currentMappingNumber++;
+    }
+
+    checkMaxNumPortMappingsReached();
+
+    return mappings;
```

Comment:
![Vulnerability](https://sonarsource.github.io/sonar-github/vulnerability.png 'Vulnerability') Vulnerability: Return a copy of "mappings". ([squid:S2384](https://sonarcloud.io/organizations/tron-zhaohong/rules#rule_key=squid%3AS2384))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=java-tron&pullRequest=1583&issues=AWZcuxrD9tx2us2yFA5d&open=AWZcuxrD9tx2us2yFA5d)

[ISSUE_KEY]: AWZcuxrD9tx2us2yFA5d